File: sol/11.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        maxval = (r - l) * min(height[l], height[r])
        maxh = max(height)
        while l + 1 < r:
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
            maxval = max(maxval, (r - l) * min(height[l], height[r]))
            if maxval >= maxh * (r - l):
                break
        return maxval

File: sol/11/11.cpp
#include <algorithm>

struct Solution {
    int maxArea(const vector<int>& height) {
        int l = 0, r = height.size() - 1;
        int maxval = (r - l) * min(height.front(), height.back()),
            maxh = *max_element(height.cbegin(), height.cend());
        while (r - l > 1){
            if (height[l] < height[r]){
                ++l;
            } else {
                --r;
            }
            maxval = max(maxval, (r - l) * min(height[l], height[r]));
            if (maxval >= maxh * (r - l)){
                break;
            }
        }
        return maxval;
    }
};
